File: Scripts/03_linear_mixed_effects_model.py
"""Linear mixed-effects model of speaking trajectory and growth mindset."""
import pandas as pd
import statsmodels.formula.api as smf

def run_lmm(data_path="Longitudinal_Data_120.csv"):
    df=pd.read_csv(data_path)
    id_var="ID" if "ID" in df.columns else ("Learner_ID" if "Learner_ID" in df.columns else df.columns[0])
    long=df.melt(id_vars=[id_var,"Growth_Mindset","Proficiency"],value_vars=["Speaking_T1","Speaking_T2","Speaking_T3"],var_name="TimePoint",value_name="Speaking_Score")
    long["Time_Num"]=long.TimePoint.map({"Speaking_T1":0,"Speaking_T2":1,"Speaking_T3":2})
    result=smf.mixedlm("Speaking_Score ~ Time_Num * Growth_Mindset + Proficiency",data=long,groups=long[id_var]).fit()
    print(result.summary())

if __name__ == "__main__": run_lmm()

File: Notebook/03_linear_mixed_effects_model.R
# Linear mixed-effects model of speaking trajectory
if (!requireNamespace("pacman", quietly=TRUE)) install.packages("pacman")
pacman::p_load(tidyverse,lme4,lmerTest,emmeans)
df <- read.csv("Longitudinal_Data_120.csv")
id_col <- if ("ID" %in% names(df)) "ID" else if ("Learner_ID" %in% names(df)) "Learner_ID" else names(df)[1]
df_long <- df %>% pivot_longer(cols=c(Speaking_T1,Speaking_T2,Speaking_T3),names_to="Time",values_to="Speaking_Score") %>% mutate(Time=factor(Time,levels=c("Speaking_T1","Speaking_T2","Speaking_T3")))
form <- as.formula(paste("Speaking_Score ~ Time * Growth_Mindset + Proficiency + (1 |",id_col,")"))
fit <- lmer(form,data=df_long); print(summary(fit))
print(emmeans(fit,~Time|Growth_Mindset,at=list(Growth_Mindset=mean(df$Growth_Mindset)+sd(df$Growth_Mindset)*c(-1,0,1))))
